Create instruments work type and pathway

Smullz622 opened the pull request:
Fixes #1586

awead commented:
@whereismyjetpack I restarted the coverage stage. Once that passes, you can merge this.

Smullz622 committed:
add new instrument work type along with related metadata fields

Smullz622 committed:
add dynamic help text for title when instrument work type is selected

Smullz622 committed:
add instrument pathway

Smullz622 committed:
add file validations checking that there is a readme as well as a separate image file

Smullz622 committed:
add keyword and publisher to instrument pathway where publisher is not visible but instead automatically set to Scholarsphere

Smullz622 committed:
yarn lint fix

jlandiseigsti commented on openapi.yml:
```diff
@@ -334,6 +334,40 @@ components:
+          example: "Classifying Independent Hybridity"
+        funding_reference:
+          type: "string"
+          example: "Classifying Independent Hybridity"
```
Do we care about these examples? If so, should they all be the same?

Smullz622 replied:
I believe we do, but I still need to touch base with Briana about what those values would actually be

Smullz622 replied:
@ajkiessl do you know if we want to have examples here? Looking further, it looks like some have them & some don't

ajkiessl replied:
I think these will be fine without examples. If anyone is using our API for uploading instruments, they'll probably know what these fields should be, or they'll at least be in close contact with us if they need to know.

jlandiseigsti commented on spec/components/citation_component_spec.rb:
```diff
@@ -56,5 +57,46 @@
+      end
+
+      context 'when there are multiple creators' do
+        let(:expected_citation) { 'Grant, Alan; Sattler, Ellie; Malcolm, Ian (2024). Citation Title [Data set]. Scholarsphere. https://doi.org/10.26207/123' }
```
There's uh, another example. See, here, now I'm sitting by myself, uh, uh, talking to myself. That's-that's chaos theory.

jlandiseigsti commented on app/controllers/dashboard/form/publish_controller.rb:
```diff
@@ -30,6 +30,7 @@ def update
           # WorkVersion#set_thumbnail_selection may be unreliable if the Shrine::ThumbnailJob is delayed
           @resource.set_thumbnail_selection
           @resource.indexing_source = Proc.new { nil }
+          @resource.set_publisher_as_scholarsphere if deposit_pathway.instrument?
```
This logic could also potentially go in the model...a `before_save` or even as part of the act of publishing.

jlandiseigsti reviewed (commented):
Looks good to me! I'll let someone more familiar with this repo give the official approve but it looks like good test coverage and good organization.

Smullz622 committed:
Merge branch 'main' into 1586-instruments-pathway

Smullz622 committed:
allow doi minting request for instrument pathway

Smullz622 committed:
remove publisher doi from visible fields since they can request to mint a doi

Smullz622 committed:
update test after removing doi from instrument details form

Smullz622 committed:
remove examples, they are not needed here

Smullz622 committed:
move setting publisher for instruments to work version model

ajkiessl commented on app/components/thumbnail_component.rb:
```diff
@@ -60,6 +60,7 @@ def icon_map
                                       dataset: 'analytics',
                                       dissertation: 'subject',
                                       image: 'image',
+                                      instrument: '',
```
You should be able to select anything from here: https://material.io/resources/icons to be the default image for instruments. Might be good to have a default, even though they're required to upload an image. I think anything that looks like a machine or tool would work.

ajkiessl commented on app/controllers/catalog_controller.rb:
```diff
@@ -142,6 +142,17 @@ def index
+    config.add_show_field 'related_identifier_tesim', label: 'Related identifier'
+    config.add_show_field 'alternative_identifier_tesim', label: 'Alternative identifier'
+    config.add_show_field 'instrument_resource_type_tesim', label: 'Instrument type'
+    config.add_show_field 'funding_reference_tesim', label: 'Funding reference'
```
To be consistent, we should make these title cased.  I actually, don't think these blacklight show fields are used anywhere.  Or they are at least overwritten here: https://github.com/psu-libraries/scholarsphere/blob/main/app/components/work_version_metadata_component.rb .  Ultimately, it looks like it's pulling labels from the en.yml: https://github.com/psu-libraries/scholarsphere/blob/main/config/locales/en.yml#L54-L81

ajkiessl commented on spec/features/dashboard/work_version_form_spec.rb:
```diff
@@ -876,6 +990,32 @@
+      let(:work_version) { work.versions.first }
+      let(:user) { work.depositor.user }
+
+      it 'shows only the fields for a scholarly work' do
```
This should be 'fields for an instrument'.

ajkiessl commented on app/views/dashboard/form/type/_work_version_fields.html.erb:
```diff
@@ -4,10 +4,20 @@
+
+  <div id="work-title-help-text" data-target="work-type.helpText" style="display: none;">
+    <p><%= t 'dashboard.form.details.instrument_title' %></p>
+  </div>
```
Adding some bootstrap classes and making the `p` element `small` will make this hint look like the one above:
```
<div class="form-text text-muted" id="work-title-help-text" data-target="work-type.helpText" style="display: none;">
    <small><%= t 'dashboard.form.details.instrument_title' %></small>
  </div>
```

ajkiessl commented on spec/features/dashboard/work_version_form_spec.rb:
```diff
@@ -1243,10 +1383,49 @@ def mock_solr_indexing_job
+      end
+    end
+  end
+
```
Since this is being done in the model now, could we remove these and just make some unit tests in the `WorkVersion` model tests?  That could save us some time in CI.

ajkiessl commented on spec/models/work_version_spec.rb:
```diff
@@ -85,6 +85,10 @@
   describe 'states' do
+    subject(:work_version) { build(:work_version, work: work) }
+
+    let(:work) { build(:work) }
```
I'm curious why this was added?

Smullz622 replied:
This was one of the weird aasm state things & part of why I was debating about where the logic should be for automatically setting publisher to scholarsphere for instruments. Originally I had it in the publish controller but didn't love that. 

I moved it to the aasm state proc on the publish state (effectively the publish action). So to transition to publish, it has to be able to check work type, but work type is delegated to work. This test was failing because work was nil, and it wasn't able to transition to publish.

ajkiessl replied:
Oh I see.  That makes sense, thanks!

ajkiessl reviewed (changes requested):
This is looking good.  Most of my requests are pretty minor.  There are some questions in there too.

Smullz622 committed:
add default thumbnail for instruments

Smullz622 committed:
a few small refactor changes

Smullz622 committed:
move spec for auto setting publisher to a unit test for better efficiency

ajkiessl commented on app/models/work_version.rb:
```diff
@@ -31,7 +31,18 @@ class WorkVersion < ApplicationRecord
+                 model: :string,
+                 instrument_type: :string,
+                 measured_variable: :string,
+                 available_date: :string,
+                 decommission_date: :string,
```
Sorry I didn't notice this before.  We should run these dates through the date validation that `published_date` runs through below:

```
validates :published_date,
            presence: true,
            edtf_date: true,
            if: :published?
```

Smullz622 replied:
In the EdtfDate module `#valid?` required the date to be a valid and present to return true so an empty available and/or decommission date would throw an error. This feels redundant because `presence: true` on `published_date` already handles this. I modified it so the module just validates format, allowing for blank values, and let `presence: true` handle that validation as needed.

ajkiessl commented on app/controllers/catalog_controller.rb:
```diff
@@ -142,6 +142,17 @@ def index
+    config.add_show_field 'related_identifier_tesim', label: 'Related Identifier'
+    config.add_show_field 'alternative_identifier_tesim', label: 'Alternative Identifier'
+    config.add_show_field 'instrument_resource_type_tesim', label: 'Resource Type'
+    config.add_show_field 'funding_reference_tesim', label: 'Funding Reference'
```
The edit and show pages actually pull labels from here: https://github.com/psu-libraries/scholarsphere/blob/main/config/locales/en.yml#L54-L81 .  We should add the fields that need to be title-cased here.

ajkiessl reviewed (changes requested):
Just two things below, but everything else looks good!

ajkiessl commented:
I just noticed this part of the open_api.yml while working on the culminating experiences stuff:

https://github.com/psu-libraries/scholarsphere/blob/70a1606f31a1b08d04f130bfb927e05818754cb3/openapi.yml#L260-L284

We'll want instruments in here too.

Smullz622 committed:
add edtf date validation for available and decommission date on instrument. Involved changing edtf module to allow a date to be valid if it is in an allowed format or blank. Other validations are checking for presence on required dates.

Smullz622 committed:
add missing fields

ajkiessl reviewed (dismissed):
LGTM

Smullz622 committed:
add help text for instrument fields

Smullz622 committed:
move hint text to be above the text field to improve accessibility

Smullz622 committed:
Merge branch 'main' into 1586-instruments-pathway

Smullz622 committed:
resolve failing specs after merging in main

Smullz622 committed:
fix js that adds help text for instrument title

Smullz622 committed:
fix spacing between hint text and fields

Smullz622 committed:
add agreement attributes to instrument pathway

Smullz622 committed:
block instruments from requesting remediation

Smullz622 committed:
niftany

Smullz622 committed:
fix a few more merge issues

Smullz622 committed:
move thumbnail help text to improve accessibility

EricDurante committed:
Customizing contributors step of deposit workflow for instruments (#1641)

* WIP on customizing contributors step of deposit workflow for instruments

This is a work in progress. At this point, I think that the workflow is
working the way that we want it to, but I haven’t written unit tests for
the new form objects yet because I’m not quite happy with the design,
and I was hoping to improve it a bit first. There are also still some
minor to-dos and potential refactorings to address.

* Refactor some behavior out of the work deposit form base class

The inheritance hierarchy of these form classes is becoming a little
more clear now that there are more of them, and the base class was
collecting too much behavior that wasn’t needed by all of its
subclasses.

* niftany

* Ci fix and fixed pathway merge issue

* Fix some tests

* Rubocop

* Fix broken tests

* Fixed form bugs

* Move contributors form builder into the deposit pathway class

* Adds unit tests for WorkDepositPathway contributor form classes

* Adds header for instruments contributors

* Fix creator building for collections bug

* Refactored/removed .build_form.  .build_creator already has the logic for when there are already creators.  So we just need the contributors form to override it in the instruments pathway.

* Adds controller logic back in to prevent creator build if creators exist

* Update no-op comment

* Delegate creators to workversion even though it wont have any

* add these back in

---------

Co-authored-by: Alex Kiessling <ajkiessl@gmail.com>

ajkiessl reviewed (approved):
LGTM
